File: game/xiangqi/logic/board.ts
import { BOARD_ALPHABET, BOARD_CELLS, BOARD_HEIGHT, BOARD_WIDTH } from './constants.gen';

export const SIDE_RED = 0;
export const SIDE_BLACK = 1;

export const EMPTY = 0;

export const KIND_GENERAL = 1;
export const KIND_ADVISOR = 2;
export const KIND_ELEPHANT = 3;
export const KIND_HORSE = 4;
export const KIND_CHARIOT = 5;
export const KIND_CANNON = 6;
export const KIND_SOLDIER = 7;

export function pieceSide(piece: number): number {
  return Math.floor((piece - 1) / 7);
}

export function pieceKind(piece: number): number {
  return ((piece - 1) % 7) + 1;
}

export function pieceFor(side: number, kind: number): number {
  return side * 7 + kind;
}

export function boardX(idx: number): number {
  return idx % BOARD_WIDTH;
}

export function boardY(idx: number): number {
  return Math.floor(idx / BOARD_WIDTH);
}

export function boardIdx(x: number, y: number): number {
  return y * BOARD_WIDTH + x;
}

export function onBoard(x: number, y: number): boolean {
  return x >= 0 && x < BOARD_WIDTH && y >= 0 && y < BOARD_HEIGHT;
}

export function flipIndex(idx: number): number {
  return BOARD_CELLS - 1 - idx;
}

export function inPalace(x: number, y: number, side: number): boolean {
  if (x < 3 || x > 5) return false;
  if (side === SIDE_RED) return y >= 0 && y <= 2;
  return y >= 7 && y <= 9;
}

export function crossedRiver(y: number, side: number): boolean {
  if (side === SIDE_RED) return y >= 5;
  return y <= 4;
}

export function forwardDir(side: number): number {
  return side === SIDE_RED ? 1 : -1;
}

export function positionKey(board: readonly number[], side: number): string {
  let key = '';
  for (let i = 0; i < BOARD_CELLS; i++) key += BOARD_ALPHABET[board[i]];
  return key + (side === SIDE_RED ? 'r' : 'b');
}

export function parseBoardRows(rows: readonly string[]): number[] {
  if (rows.length !== BOARD_HEIGHT) throw new RangeError('invalid board rows');
  const board = new Array<number>(BOARD_CELLS).fill(EMPTY);
  for (let y = 0; y < BOARD_HEIGHT; y++) {
    const row = rows[y];
    if (row.length !== BOARD_WIDTH) throw new RangeError('invalid board row length');
    for (let x = 0; x < BOARD_WIDTH; x++) {
      const code = BOARD_ALPHABET.indexOf(row[x]);
      if (code < 0) throw new RangeError('invalid board character');
      board[boardIdx(x, y)] = code;
    }
  }
  return board;
}

File: server/internal/game/games/xiangqi/board.go
package xiangqi

import (
	"errors"
	"strings"
)

const (
	GameID      = "xiangqi"
	TurnSeconds = 60

	BoardWidth  = 9
	BoardHeight = 10
	BoardCells  = BoardWidth * BoardHeight

	HalfmoveDrawPlies = 120
	RepetitionLimit   = 3

	SideRed   = 0
	SideBlack = 1

	Empty = 0

	KindGeneral  = 1
	KindAdvisor  = 2
	KindElephant = 3
	KindHorse    = 4
	KindChariot  = 5
	KindCannon   = 6
	KindSoldier  = 7

	MaxPiece = 14
)

const BoardAlphabet = ".KAEHRCPkaehrcp"

var startRows = [BoardHeight]string{
	"RHEAKAEHR",
	".........",
	".C.....C.",
	"P.P.P.P.P",
	".........",
	".........",
	"p.p.p.p.p",
	".c.....c.",
	".........",
	"rheakaehr",
}

var startBoard = mustParseBoard(startRows)

func pieceSide(p int) int { return (p - 1) / 7 }

func pieceKind(p int) int { return (p-1)%7 + 1 }

func pieceFor(side, kind int) int { return side*7 + kind }

func boardX(idx int) int { return idx % BoardWidth }

func boardY(idx int) int { return idx / BoardWidth }

func boardIdx(x, y int) int { return y*BoardWidth + x }

func onBoard(x, y int) bool {
	return x >= 0 && x < BoardWidth && y >= 0 && y < BoardHeight
}

func inPalace(x, y, side int) bool {
	if x < 3 || x > 5 {
		return false
	}
	if side == SideRed {
		return y >= 0 && y <= 2
	}
	return y >= 7 && y <= 9
}

func crossedRiver(y, side int) bool {
	if side == SideRed {
		return y >= 5
	}
	return y <= 4
}

func forwardDir(side int) int {
	if side == SideRed {
		return 1
	}
	return -1
}

func validAdvisorSpot(x, y, side int) bool {
	if !inPalace(x, y, side) {
		return false
	}
	if side == SideRed {
		return (x+y)%2 == 1
	}
	return (x+y)%2 == 0
}

func validElephantSpot(x, y, side int) bool {
	ey := y
	if side == SideBlack {
		ey = BoardHeight - 1 - y
	}
	switch ey {
	case 0, 4:
		return x == 2 || x == 6
	case 2:
		return x%4 == 0
	}
	return false
}

func validSoldierSpot(x, y, side int) bool {
	ey := y
	if side == SideBlack {
		ey = BoardHeight - 1 - y
	}
	if ey < 3 {
		return false
	}
	if ey <= 4 {
		return x%2 == 0
	}
	return true
}

func parseBoardRows(rows [BoardHeight]string) ([]int, error) {
	board := make([]int, BoardCells)
	for y, row := range rows {
		if len(row) != BoardWidth {
			return nil, errors.New("invalid board row length")
		}
		for x := 0; x < BoardWidth; x++ {
			code := strings.IndexByte(BoardAlphabet, row[x])
			if code < 0 {
				return nil, errors.New("invalid board character")
			}
			board[boardIdx(x, y)] = code
		}
	}
	return board, nil
}

func mustParseBoard(rows [BoardHeight]string) []int {
	board, err := parseBoardRows(rows)
	if err != nil {
		panic(err)
	}
	return board
}

func formatBoard(board []int) [BoardHeight]string {
	var rows [BoardHeight]string
	for y := 0; y < BoardHeight; y++ {
		var b strings.Builder
		for x := 0; x < BoardWidth; x++ {
			b.WriteByte(BoardAlphabet[board[boardIdx(x, y)]])
		}
		rows[y] = b.String()
	}
	return rows
}

func positionKey(board []int, side int) string {
	var b strings.Builder
	b.Grow(BoardCells + 1)
	for _, p := range board {
		b.WriteByte(BoardAlphabet[p])
	}
	if side == SideRed {
		b.WriteByte('r')
	} else {
		b.WriteByte('b')
	}
	return b.String()
}
